main.cpp

```diff
@@ -103,246 +106,333 @@ good to go!
 
+struct DoubleType
+{
+    DoubleType(double nuDouble)
+    {
+        value = new double(nuDouble);
+    }
+    ~DoubleType()
+    {
+        delete value;
+        value = nullptr;
+    }
+    double* value = nullptr;
+
+    DoubleType& add( double rhs );
+    DoubleType& subtract( double rhs );
+    DoubleType& multiply( double rhs );
+    DoubleType& divide( double rhs );
+
+    DoubleType& add( const FloatType& rhs );
+    DoubleType& subtract( const FloatType& rhs );
+    DoubleType& multiply( const FloatType& rhs );
+    DoubleType& divide( const FloatType& rhs );
+
+    DoubleType& add( const IntType& rhs );
+    DoubleType& subtract( const IntType& rhs );
+    DoubleType& multiply( const IntType& rhs );
+    DoubleType& divide( const IntType& rhs );
+};
 
-struct FloatType
+struct IntType
 {
-    float add( float lhs, float rhs );
-    float subtract( float lhs, float rhs );
-    float multiply( float lhs, float rhs );
-    float divide( float lhs, float rhs );
+    IntType(int nuInt)
+    {
+        value = new int(nuInt);
+    }
+    ~IntType()
+    {
+        delete value;
+        value = nullptr;
+    }
+    int* value = nullptr;
+
+    IntType& add( int rhs );
+    IntType& subtract( int rhs );
+    IntType& multiply( int rhs );
+    IntType& divide( int rhs );
+
+    IntType& add( const DoubleType& );
+    IntType& subtract( const DoubleType& );
+    IntType& multiply( const DoubleType& );
+    IntType& divide( const DoubleType& );
+
+    IntType& add( const FloatType& );
+    IntType& subtract( const FloatType& );
+    IntType& multiply( const FloatType& );
+    IntType& divide( const FloatType& );
 };
 
-float FloatType::add( float lhs, float rhs )
+//========================================================
+
+FloatType& FloatType::add( float rhs )
 {
-    return (lhs + rhs);
+    *this->value += rhs;
+    return *this;
 }
-float FloatType::subtract( float lhs, float rhs )
+FloatType& FloatType::subtract( float rhs )
 {
-    return (lhs - rhs);
+    *this->value -= rhs;
+    return *this;
 }
-float FloatType::multiply( float lhs, float rhs )
+FloatType& FloatType::multiply( float rhs )
 {
-    return (lhs * rhs);
+    *this->value *= rhs;
+    return *this;
 }
-float FloatType::divide( float lhs, float rhs )
+FloatType& FloatType::divide( float rhs )
 {
     if ( rhs == 0.0f )
-        std::cout << "warning, floating point division by zero returns 'inf' !" << std::endl;
+        std::cout << "warning: floating point division by zero!" << std::endl;
 
-    return (lhs / rhs);
+    *this->value /= rhs;
+    return *this;
 }
 
+FloatType& FloatType::add( const DoubleType& rhs )
+{
+    return add( float(*rhs.value) );
+}
+FloatType& FloatType::subtract( const DoubleType& rhs )
+{
+    return subtract( float(*rhs.value) );
+}
+FloatType& FloatType::multiply( const DoubleType& rhs )
+{
+    return multiply( float(*rhs.value) );
+}
+FloatType& FloatType::divide( const DoubleType& rhs )
+{
+    return divide( float(*rhs.value) );
+}
```

Comment:
This casting is premature.
it is also C-style casting, which we don't use when writing C++
remove the casting so your code produces the 18 warnings, as shown in the expected output.
you'll learn to cast properly in the next project part.
- [ ] revise